src/aws/flb_aws_credentials.c

```diff
@@ -323,6 +324,183 @@ struct flb_aws_provider *flb_standard_chain_provider_create(struct flb_config
+        goto error;
+    }
+
+    region = flb_output_get_property(config_key_region, ins);
+    if (!region) {
+        flb_plg_error(ins, "aws_auth enabled but %s not set", config_key_region);
+        goto error;
+    }
+
+    /* Use null sts_endpoint if none provided */
+    sts_endpoint = flb_output_get_property(config_key_sts_endpoint, ins);
+
+    aws_provider = flb_standard_chain_provider_create(config,
+                                                      cred_tls,
+                                                      (char *) region,
+                                                      (char *) sts_endpoint,
+                                                      NULL,
+                                                      flb_aws_client_generator());
+    if (!aws_provider) {
+        flb_plg_error(ins, "Failed to create AWS Credential Provider");
+        goto error;
+    }
+
+    role_arn = flb_output_get_property(config_key_role_arn, ins);
+    if (role_arn) {
+        /* Use the STS Provider */
+        base_aws_provider = aws_provider;
+        external_id = flb_output_get_property(config_key_external_id, ins);
+
+        session_name = flb_sts_session_name();
+        if (!session_name) {
+            flb_plg_error(ins, "Failed to create aws iam role "
+                        "session name");
+            goto error;
+        }
+
+        /* STS provider needs yet another separate TLS instance */
+        sts_tls = flb_tls_create(FLB_TRUE,
+                                            ins->tls_debug,
+                                            ins->tls_vhost,
+                                            ins->tls_ca_path,
+                                            ins->tls_ca_file,
+                                            ins->tls_crt_file,
+                                            ins->tls_key_file,
+                                            ins->tls_key_passwd);
+        if (!sts_tls) {
+            flb_errno();
+            goto error;
+        }
+
+        aws_provider = flb_sts_provider_create(config,
+                                                    sts_tls,
+                                                    base_aws_provider,
+                                                    (char *) external_id,
+                                                    (char *) role_arn,
+                                                    session_name,
+                                                    (char *) region,
+                                                    (char *) sts_endpoint,
+                                                    NULL,
+                                                    flb_aws_client_generator());
+        if (!aws_provider) {
+            flb_plg_error(ins, "Failed to create AWS STS Credential "
+                        "Provider");
+            goto error;
+        }
+    }
+
+    /* initialize credentials in sync mode */
+    aws_provider->provider_vtable->sync(aws_provider);
```

Comment:
could you explain why it need to be set as sync at first and set back to async quickly?

Reply:
The way i understand it is that during plugin init, there is no such thing as async networking (this might have changed recently). That is because setup occurs before the coroutines are allowed to swap out. Really the reason why I did this was because everywhere else we setup the cred provider, we do this.

Reply:
Yea I added this a while ago to fix a bug... AFAIK the init callbacks still must run in sync mode.